source/m3_api_uvwasi.c

```diff
@@ -0,0 +1,410 @@
+    m3ApiGetArgMem   (char*                , argv_buf)
+
+    if (runtime == NULL) { m3ApiReturn(__WASI_EINVAL); }
+
+    for (u32 i = 0; i < runtime->argc; ++i)
+    {
+        argv[i] = m3ApiPtrToOffset (argv_buf);
+
+        size_t len = strlen (runtime->argv [i]);
+        memcpy (argv_buf, runtime->argv [i], len);
+        argv_buf += len;
+        * argv_buf++ = 0;
+    }
+
+    m3ApiReturn(UVWASI_ESUCCESS);
+}
+
+m3ApiRawFunction(m3_wasi_unstable_args_sizes_get)
+{
+    m3ApiReturnType  (uint32_t)
+    m3ApiGetArgMem   (__wasi_size_t*       , argc)
+    m3ApiGetArgMem   (__wasi_size_t*       , argv_buf_size)
+
+    if (runtime == NULL) { m3ApiReturn(__WASI_EINVAL); }
+
+    *argc = runtime->argc;
+    *argv_buf_size = 0;
+    for (u32 i = 0; i < runtime->argc; ++i)
+    {
+        * argv_buf_size += strlen (runtime->argv [i]) + 1;
+    }
+    m3ApiReturn(UVWASI_ESUCCESS);
+}
+
+m3ApiRawFunction(m3_wasi_unstable_environ_get)
+{
+    m3ApiReturnType  (uint32_t)
+    m3ApiGetArgMem   (u32*                 , env)
+    m3ApiGetArgMem   (char*                , env_buf)
+
+    char **environment;
+    uvwasi_errno_t err;
+
+    environment = calloc(uvwasi.envc, sizeof(char *));
+    if (environment == NULL) {
+        m3ApiReturn(UVWASI_ENOMEM);
+    }
+
+    err = uvwasi_environ_get(&uvwasi, environment, env_buf);
+    if (err != UVWASI_ESUCCESS) {
+        free(environment);
+        m3ApiReturn(err);
+    }
+
+    uint32_t environ_buf_offset = m3ApiPtrToOffset(env_buf);
+
+    for (u32 i = 0; i < uvwasi.envc; ++i)
+    {
+        uint32_t offset = environ_buf_offset +
+                          (environment[i] - environment[0]);
+        env[i] = offset;
+    }
+
+    free(environment);
+    m3ApiReturn(UVWASI_ESUCCESS);
+}
+
+m3ApiRawFunction(m3_wasi_unstable_environ_sizes_get)
+{
+    m3ApiReturnType  (uint32_t)
+    m3ApiGetArgMem   (__wasi_size_t*       , env_count)
+    m3ApiGetArgMem   (__wasi_size_t*       , env_buf_size)
+
+    size_t count;
+    size_t buf_size;
+    uvwasi_errno_t err;
+
+    err = uvwasi_environ_sizes_get(&uvwasi, &count, &buf_size);
+    *env_count = count;
+    *env_buf_size = buf_size;
+    m3ApiReturn(err);
+}
+
+m3ApiRawFunction(m3_wasi_unstable_fd_prestat_dir_name)
+{
+    m3ApiReturnType  (uint32_t)
+    m3ApiGetArg      (uvwasi_fd_t          , fd)
+    m3ApiGetArgMem   (char*                , path)
+    m3ApiGetArg      (size_t               , path_len)
+
+    m3ApiReturn(uvwasi_fd_prestat_dir_name(&uvwasi, fd, path, path_len));
+}
+
+m3ApiRawFunction(m3_wasi_unstable_fd_prestat_get)
+{
+    m3ApiReturnType  (uint32_t)
+    m3ApiGetArg      (uvwasi_fd_t          , fd)
+    m3ApiGetArgMem   (uint32_t *           , buf)
+
+    uvwasi_prestat_t prestat;
+    uvwasi_errno_t err;
+
+    err = uvwasi_fd_prestat_get(&uvwasi, fd, &prestat);
+    if (err != UVWASI_ESUCCESS) {
+        m3ApiReturn(err);
+    }
+
+    // TODO(cjihrig): This memory writing logic is wrong.
```

Comment:
BTW,  any ideas of how it should look like?

Reply:
I believe the `pr_type` field should be written to the first four bytes in `buf`, and the `u.dir.pr_name_len` field should be written to the next four bytes. At least that's what we've done in [Node](https://github.com/nodejs/node/blob/09a50d3c5a19c0874cd127b2e55be500ea25372b/src/node_wasi.cc#L694-L695).

Reply:
In this case, it should be `*(buf + 1)`

Reply:
Yes, this should be `*(buf + 1)`, and the last parameter to `m3_wasi_unstable_fd_prestat_dir_name()` should be a `__wasi_size_t` instead of a `size_t`. When I make that change, the data looks correct, but the test starts to fail when trying to open the file being passed as the `cat` argument.